indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java

```diff
@@ -529,12 +494,55 @@ private FiniteAppenderatorDriver newDriver(
         new NoopSegmentHandoffNotifierFactory(), // don't wait for handoff since we don't serve queries
         new ActionBasedUsedSegmentChecker(toolbox.getTaskActionClient()),
         toolbox.getObjectMapper(),
-        Integer.MAX_VALUE, // rows for a partition is already determined by the shardSpec
+        // If targetPartitionSize is null, numShards must be set which means intervals are already partitioned with
+        // proper shardSpecs. See determineShardSpecs().
+        tuningConfig.getTargetPartitionSize() == null ? Integer.MAX_VALUE : tuningConfig.getTargetPartitionSize(),
+        tuningConfig.getMaxPersistedSegmentsBytes(),
         0,
         metrics
     );
   }
 
+  /**
+   * This interface represents a map of (Interval, ShardSpec) and is used for easy shardSpec generation.  The most
+   * important method is {@link #updateShardSpec(Interval)} which updates the map according to the type of shardSpec.
+   */
+  private interface ShardSpecs
+  {
+    /**
+     * Return the key set of the underlying map.
+     *
+     * @return a set of intervals
+     */
+    Collection<Interval> getIntervals();
+
+    /**
+     * Indicate that the type of shardSpecs is extendable like {@link LinearShardSpec}.
+     *
+     * @return true if the type of shardSpecs is extendable
+     */
+    boolean isExtendable();
+
+    /**
+     * Return a shardSpec for the given interval, timestamp and input row.
+     *
+     * @param interval  interval for shardSpec
+     * @param timestamp timestamp of input row
+     * @param row       input row
+     * @return a shardSpec
+     */
+    ShardSpec getShardSpec(Interval interval, long timestamp, InputRow row);
```

Comment:
Do you foresee any situations where the timestamp here would be different from the row timestamp?

If not, maybe the timestamp parameter could be dropped and read from the row instead.